### PR TITLE
#472 - Removed # from being required in IDs + restricted IDs to Base64URL char set

### PR DESCRIPTION
**Pull request by thehenrytsai**
1. Removed # from being required/allowed in IDs.
1. Added '#' to IDs during external DID document transform.
1. Added base64url character set check for IDs.

### CODE REVIEW

File: lib/core/versions/latest/DocumentComposer.ts [hunk 1]
```diff
@@ -249,15 +248,28 @@ export default class DocumentComposer {
     }
   }
 
-  private static validateServiceEndpointId (serviceEndpointId: any) {
-    if (typeof serviceEndpointId !== 'string') {
-      throw new SidetreeError(ErrorCode.DocumentComposerPatchServiceEndpointIdNotString);
+  private static validateId (id: any) {
+    if (typeof id !== 'string') {
+      throw new SidetreeError(ErrorCode.DocumentComposerIdNotString);
```
Comment (mudiali) [inline lib/core/versions/latest/DocumentComposer.ts]: Nit: include the input in the error message? The tests will have to change to use the JasmieSidetreeErrorValidator class though so up to you if the change is big.

File: lib/core/versions/latest/DocumentComposer.ts [hunk 2]
```diff
@@ -381,11 +389,19 @@ export default class DocumentComposer {
 
     didDocument.id = did;
 
-    // Only update the publickey if the array is present
+    // Only update `publickey` if the array is present
```
Comment (OR13) [inline lib/core/versions/latest/DocumentComposer.ts]: You could include jsonld.js as a dependency, and use `@base` to do this, for all the properties at once... `assertionMethod` (required by the VC Data Model), `authentication` (required by CHAPI), keyAgreementKey, (required by EDVs)...

This solution looks a little brittle to me, but it gets the job done for the current set of features that ION DIDs support.

File: lib/core/versions/latest/DocumentComposer.ts (hunk 1 above)
Comment (thehenrytsai) [reply lib/core/versions/latest/DocumentComposer.ts]: Updated.

File: lib/core/versions/latest/DocumentComposer.ts (hunk 2 above)
Comment (thehenrytsai) [reply lib/core/versions/latest/DocumentComposer.ts]: Created issue #512 to rework the transform logic.

**Pull request merged**